# SQL Queries and AACT Data

In this assignment, you will answer questions about clinical trials registered with the US government. The data is a subset of the data made available through the Clinical Trials Transformation Initiative. 
The data provided is a subset of the full data set, so be sure to use the
version of the data available on the course Canvas site.  If you are
interested, you can learn more about the dataset at
http://aact.ctti-clinicaltrials.org.


## What's In and Out of Scope

This is intended to be a declarative SQL query assignment. Therefore, you must write
queries in SQL (not functions, and not in Python).  You
may use VIEWs as needed and you may use standard built-in PostgreSQL
functions (e.g. ROUND or CASE statements). If you're not sure if
something is allowed, ask!

## Getting Started

First,  go to your database, and create the tables found in the tablesA2.sql file.

## Load the data

Load the data needed for the assignment. You should do this in
pgAdmin. The files are provided in the Canvas assignment. They are .sql files. You can copy & paste the contents into a pgAdmin window and run them.


## Cautions
You must use the table and attribute names provided. Do not rename anything. We want to be able to run your submission and see all of the  results be generated.


At the end of your submission, include **DROP** or **DELETE** statements that remove any VIEWs or TABLEs your code created. Points will be deducted if you do not include these statements. Basically, you need to "clean up" after your code.

## Turnin
Fill in this notebook and download it as a notebook (.ipynb).  By 11:55P on the due date, submit this document electronically to Canvas.   Answers in the comments section of CANVAS are not acceptable and will not be read.  Be sure to turn in the file with your answers in it, not just the empty assignment notebook


## Grading
 The number of points for each query is indicated in the question.   
If you don't get the right answer or your code is not correct, you won't get all of the points; partial credit may be given
at the discretion of the grader.

## Academic Honesty

The following level of collaboration is allowed on this assignment: You
may discuss the assignment with your classmates at a high level. Any
issues getting Postgres running is totally fine. What
is not allowed is direct examination of anyone else's SQL code (on a
computer, email, whiteboard, etc.) or allowing anyone else to see your
SQL code. You may not discuss query results with your classmates.

You may use the search engine of your choice to lookup the syntax for
SQL commands, but may not use it to find answers to queries.



## Connect to your database



In [1]:
# set this value to your database's username
dbuser = "jf58"
# set this value to your database's database name
dbName = "jf58db"
# set this value to your database's endpoint
endpoint = "ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com"

Run the next cell to enter your password 

In [2]:
import getpass
# enter the password to your database
password = getpass.getpass()


········


In [3]:
# build the connection string
def make_conn_str(dbuser, password, endpoint, dbName):

    return f"postgresql+psycopg2://{dbuser}:{password}@{endpoint}/{dbName}"


In [4]:
conn_str = make_conn_str(dbuser, password,endpoint, dbName)

In [5]:
conn_str

'postgresql+psycopg2://jf58:13775082158Fj@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db'

In [6]:
%load_ext sql

In [7]:
%sql $conn_str

'Connected: jf58@jf58db'

Check connectivity by running the next cell. The query should the date and time where the server is located.

In [8]:
%%sql 
SELECT NOW();

 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
1 rows affected.


[(datetime.datetime(2020, 2, 13, 5, 15, 53, 391510, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)),)]

Limit queries to 100 results. Increase this value if needed, but recognize that your JN will increase in size as well. 

In [9]:
%config SqlMagic.displaylimit=100

###  Create the database tables and load the data into the database

* Download the zip file containing the data from Canvas
* Unzip the file
* Start up pgAdmin
* Navigate to your database
* Select Tools/Query Tool from the top menu bar
* Open the file named ```A2S20tabledefs.sql```
* Run the file (using the play button or lightening bolt, depending on your version of pgAdmin)
* Verify that the tables have been created:
    * In the main pgAdmin window, navigate to Schemas/public/Tables
    * Right click and select Refresh
    * Verify that the new tables are there
* In turn, open and run each of the following files:
  
    ```responsible_parties.sql``` (took about 1 minute on my database)
    
    ```overall_officials.sql``` (took about 1 minute on my database)
    
    ```sponsors.sql``` (took about 1 and a half minutes on my database)
    
    ```factor.sql``` (took a few seconds)
         
    ```studies.sql``` (took a few seconds)
    
    ```conditions.sql``` (took a few seconds)
    

# Queries
Answer all of the questions below by writing and executing SQL queries. The queries must contain ONLY the answer to the question (no extra rows or columns). You may need to explore the database a bit prior to generating your final solutions. 

1 (8 points) We want a master list of the names of all the people in charge in the database.  Write queries to fill in the values in the following table  

|                     | # Names | # Unique Names |
|---------------------|---------|----------------|
| responsible_parties | 1       | 2              |
| overall_officials   | 3       | 4              |
| sponsors            | 5       | 6              |
| total               | 7       | 8              |


In [10]:
%%sql
-- query for entry 1
SELECT COUNT(name)
FROM responsible_parties

 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
1 rows affected.


[(112214,)]

In [11]:
%%sql
-- query for entry 2
SELECT COUNT(DISTINCT name)
FROM responsible_parties

 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
1 rows affected.


[(61334,)]

In [12]:
%%sql
-- query for entry 3
SELECT COUNT(name)
FROM overall_officials

 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
1 rows affected.


[(276769,)]

In [13]:
%%sql
-- query for entry 4
SELECT COUNT(DISTINCT name)
FROM overall_officials

 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
1 rows affected.


[(180010,)]

In [14]:
%%sql
-- query for entry 5
SELECT COUNT(name)
FROM sponsors

 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
1 rows affected.


[(411461,)]

In [15]:
%%sql
-- query for entry 6
SELECT COUNT(DISTINCT name)
FROM sponsors

 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
1 rows affected.


[(47480,)]

In [16]:
%%sql
-- query for entry 7
SELECT 
(SELECT COUNT(name)
FROM responsible_parties)+
(SELECT COUNT(name)
FROM overall_officials)+
(SELECT COUNT(name)
FROM sponsors) AS "RESULT"

 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
1 rows affected.


[(800444,)]

In [17]:
%%sql
-- query for entry 8
SELECT 
(SELECT COUNT(DISTINCT name)
FROM responsible_parties)+
(SELECT COUNT(DISTINCT name)
FROM overall_officials)+
(SELECT COUNT(DISTINCT name)
FROM sponsors) AS "RESULT"

 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
1 rows affected.


[(288824,)]

Replace the numbers in this table with your query results

|                     | # Names | # Unique Names |
|---------------------|---------|----------------|
| responsible_parties | 112214  | 61334          |
| overall_officials   | 276769  | 180010         |
| sponsors            | 411461  | 47480          |
| total               | 800444  | 288824         |

9 (2 points) Are there still duplicate entries in the table? If there are, modify query 8 to remove them. Include your query and the final count of the number of unique names across the 3 tables.



your text answer here

In [18]:
%%sql
DROP TABLE IF EXISTS a21 CASCADE;
CREATE TABLE a21(
    name character varying
);
INSERT INTO a21 SELECT DISTINCT name FROM responsible_parties;
INSERT INTO a21 SELECT DISTINCT name FROM overall_officials;
INSERT INTO a21 SELECT DISTINCT name FROM sponsors;
SELECT COUNT(DISTINCT name) FROM a21;

 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
Done.
Done.
61335 rows affected.
180010 rows affected.
47480 rows affected.
1 rows affected.


[(275386,)]

10 (4 points) Sort the results of the previous query and examine the results. 

Note you might not want to look at ALL the data at one time, as it can overload the server, and is just too much to review.

Give 2 different examples of names that are likely to be duplicates, but are showing up as separate entries. The reason why they are showing up as duplicates must be different for the two examples you provide. Provide the reason why you believe the names are the same.


In [19]:
%%sql
SELECT DISTINCT name FROM  overall_officials

 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
180010 rows affected.


[('/ / /, /',),
 ('. .',),
 ('. ., .',),
 ('01 Studienregister MasterAdmins',),
 ('02915 02915, Ph.D.',),
 ('1-317-615-4559 Mon - Fri 9 AM - 5 PM Eastern time (UTC/GMT - 5 hours, EST)',),
 ('1-858-255-5959 Mon-Fri from 9 AM to 5 PM Pacific Time (PST)',),
 ('1-877-CTLILLY (1-877-285-4559) or 1-317-615-4559 Mon - Fri 9 AM - 5 PM Eastern time (UTC/GMT - 5 hours, EST)',),
 ('1-877-CTLILLY (1-877-285-4559) or 1-317-615-4559 Mon-Fri 9 AM - 5 PM Eastern time (UTC/GMT - 5hours, EST)',),
 ('4195 Extraction Study Team',),
 ('4195 LV Lead Chronic Performance Study Team',),
 ('4196 LV Lead Chronic Performance Study Team',),
 ('4965 Post-Approval Study Clinical Trial Leader',),
 ('4968 Clinical Trial Leader',),
 ('86-20-81888553 Zhang, M.D&Ph.D',),
 ('90 212 4404000 90 212 4404000, MD',),
 ('A A. Franken, MD PhD',),
 ('Aage Indahl, Phd',),
 ('Aage Indahl, Ph.d',),
 ('Aage Telnes, MD',),
 ('A Ahsan Ejaz, M.D',),
 ('A. Ahsan Ejaz, MD',),
 ('Aala Shalaby, MD, FACC',),
 ('AA Levine',),
 ('A. Algra, Professor',),
 ('Aali M Shah, MD',),
 ('Aalim Weljie, PhD',),
 ('Aali Sheen',),
 ('Aalo Bistritz, MD',),
 ('Aalok V Agarwala, MD, MBA',),
 ('Aalt Bast, PhD',),
 ('Aalt Bast, PhD, Prof.',),
 ('Aalt Bast, Prof. Dr.',),
 ('Aalt Bast, Prof.Dr.',),
 ('Aalt Bast, Professor',),
 ('Aalt Bast, Prof, PhD',),
 ('A Al-Zahrani, M.D.',),
 ('A.A.M. Franken, MD, PhD',),
 ('A Amiel, Dr',),
 ('Aamina M Khan',),
 ('Aamir J. Khan, MD/PhD.',),
 ('Aamir Khan',),
 ('aamir siddiqui, md',),
 ('aamir siddiqui, MD',),
 ('Aamir Siddiqui, MD',),
 ('A.A.M. Masclee',),
 ('A.A.M. Masclee, MD, PhD',),
 ('A AM Masclee, Prof. Dr.',),
 ('A.A.M. Masclee, Professor',),
 ('A.A.M. Masclee, Prof. MD. PhD',),
 ('A.A.M. van der Veldt',),
 ('Aanand D Naik, MD',),
 ('Aanand D Naik, MD BA',),
 ('Aapo Tahkola, LM,PHD st.',),
 ('A. Ardizzoia, MD',),
 ('Aarjan P Snoek, MBChB, FRCA',),
 ('Aarne Feldheiser, MD, PhD',),
 ('Aaron A Bare, MD',),
 ('Aaron A Cohen, MD, MSc',),
 ('Aaron Allen, MD',),
 ('Aaron Allen, M.D.',),
 ('Aaron A Reeves, M.D.',),
 ('Aaron B Caughey, MD,PhD',),
 ('Aaron Berman, MD',),
 ('Aaron B Holley, MD',),
 ('Aaron Biesbrock, DMD, PhD',),
 ('Aaron Boster, MD',),
 ('Aaron Brody, MD, MPH',),
 ('Aaron Burstein, Pharm.D.',),
 ('Aaron B Waxman, MD PhD',),
 ('Aaron Calodney, MD',),
 ('Aaron Carrel, MD',),
 ('Aaron Carrel, MD, MPH',),
 ('Aaron Carr, MD',),
 ('Aaron Chamberlain, MD',),
 ('Aaron Chiu, MD',),
 ('Aaron C Lentz, MD',),
 ('Aaron Cumpston, PharmD',),
 ('Aaron Cumpston, PharmD, BCOP',),
 ('Aaron D Fox, MD, MS',),
 ('Aaron Donoghue, MD',),
 ('Aaron Donoghue, MD, MSCE',),
 ('Aaron E Bair, MD',),
 ('Aaron E Carroll, MD, MS',),
 ('Aaron E Chen, MD',),
 ('Aaron E Katz, MD',),
 ('Aaron Ellenbogen, DO',),
 ('Aaron F Gomes, MDS',),
 ('Aaron Fischman, MD',),
 ('Aaron Fischman, M.D',),
 ('Aaron Folsom',),
 ('Aaron Folsom, MD, MPH',),
 ('Aaron Fox, MD, MS',),
 ('Aaron Fox, PhD',),
 ('Aaron Friedman, MD',),
 ('Aaron Frimerman, MD',),
 ('Aaron Frimerman, Professor',),
 ('Aaron Gillam',),
 ('Aaron Grantham, MD',),
 ('Aaron Hamilton, MD',),
 ('Aaron Hamvas, MD',),
 ('Aaron Hansen, M.D.',),
 ('AARON HANSEN, M.D.',),
 ('Aaron H Burstein, PharmD',),
 ('Aaron Hogue, PhD',),
 ('Aaron Hogue, Ph.D.',),
 ('Aaron H Wolfson, MD',),
 ('Aaron H. Wolfson, MD',),
 ('Aaron I Baum, PhD',),
 ('Aaron I Vinik, MD, PhD',),
 ('Aaron I Vinik, PhD, MD',),
 ('Aaron Jacobs',),
 ('Aaron J Blashill, PhD',),
 ('Aaron J Krych',),
 ('Aaron J Newman, PhD',),
 ('Aaron Johnson, MD',),
 ('Aaron Keshen, MD, FRCPC',),
 ('Aaron Lazorwitz, MD',),
 ('Aaron L Boster, MD',),
 ('Aaron L Carrel',),
 ('Aaron L Ellenbogen, DO, MPH',),
 ('Aaron Mangold',),
 ('Aaron Mangold, MD',),
 ('Aaron Mansfield',),
 ('Aaron Mansfield, MD',),
 ('Aaron M Burnett, MD',),
 ('Aaron M Cypess, M.D.',),
 ('Aaron M Cypess, MD PhD',),
 ('Aaron Michels, MD',),
 ('Aaron Miller, MD',),
 ('Aaron Milstone, MD MHS',),
 ('Aaron Milstone, MD, MHS',),
 ('Aaron M Joffe, DO',),
 ('Aaron M Johnson, PhD',),
 ('Aaron Moberly, MD',),
 ('Aaron Nauth, MD FRCS(C)',),
 ('Aaron 

your text answer here
Example1: 
aamir siddiqui, MD
Aamir Siddiqui, MD
Reason:Both of them have the same name and degree. The difference is that the first letter is capitalized and the other is not.
Example2:
Aaron Allen, MD
Aaron Allen, M.D.
Reason:The spelling of their names and their degrees are the same, only the abbreviations of the degrees are different.

11 (4 points) Make 2 suggestions for how to change the data storage structure to reduce the number of these remaining duplicates. 



your text answer here
1.Standardize the case of names (such as uniform capitalization)
1.Standardize the input format of name and degree, such as no more spaces, commas and other symbols

Studies have shown that lifestyle choices can have a big impact on longevity. Let's investigate which studies deal with lifestyle aspects. In particular, we will look at conditions pertaining to 
* alcohol use
* diet
* lifestyle
* nutrition
* sexual activity
* substance abuse
* tobacco use
* weight

These factors are all in the ```factor``` table. This table contains two attributes: ```category``` and ```condition```. The conditions are drawn from the ```conditions table```, and are grouped into the categories mentioned above.

Let's compute a score that incorporates these factors for each study.

12 (2 points) How many studies have conditions that fall into the category of 'nutrition'?

In [20]:
%%sql
DROP TABLE IF EXISTS a22 CASCADE;
CREATE TABLE a22(
    category varchar(200),
    condition varchar(200)
);
INSERT INTO a22 SELECT category,condition FROM factor WHERE category = 'nutrition';
SELECT COUNT (DISTINCT s.nct_id)
FROM a22 AS f JOIN conditions AS s ON f.condition = s.name;

 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
Done.
Done.
193 rows affected.
1 rows affected.


[(49,)]

13 (10 points) Compute a risk factor score that is just the total number of factors covered by the study, ignoring the factor category completely. List the nct_ids and counts of the factors for every study that has at least 6 factors. Sort by nct_id.

In [20]:
%%sql
DROP TABLE IF EXISTS a23 CASCADE;
CREATE TABLE a23(
    nct_id CHARACTER(11),
    count integer
);
INSERT INTO a23
SELECT nct_id, COUNT(category)
FROM factor AS s  JOIN conditions AS f ON s.condition = f.name
GROUP BY nct_id
ORDER BY nct_id;

SELECT nct_id,count FROM a23
WHERE count>=6


 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
Done.
Done.
376 rows affected.
1 rows affected.


[('NCT02914002', 6)]

14 (10 points) Compute a category score  for each study based on the number of categories covered. Add 1 to the study score if the study has ANY condition listed in the factor table for each category.  Each category can be counted at most once.

How many studies have category scores of at least 3?

In [22]:
%%sql 
DROP TABLE IF EXISTS a24 CASCADE;
CREATE TABLE a24(
    nct_id CHARACTER(11),
    count integer
);
INSERT INTO a24
SELECT m.nct_id,COUNT(DISTINCT category)
FROM factor AS s JOIN conditions AS f on s.condition = f.name RIGHT OUTER JOIN studies AS m ON f.nct_id = m.nct_id
GROUP BY m.nct_id
ORDER BY m.nct_id;

SELECT COUNT(DISTINCT nct_id) FROM a24
WHERE count>=3;

 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
Done.
Done.
11977 rows affected.
1 rows affected.


[(3,)]

15 (10 points) How many studies in our database cover none of the conditions in the factor table? 


In [23]:
%%sql
DROP TABLE IF EXISTS a25 CASCADE;
CREATE TABLE a25(
    nct_id CHARACTER(11),
    count integer
);
INSERT INTO a25
SELECT m.nct_id,COUNT(DISTINCT category)
FROM factor AS s JOIN conditions AS f on s.condition = f.name RIGHT OUTER JOIN studies AS m ON f.nct_id = m.nct_id
GROUP BY m.nct_id
ORDER BY m.nct_id;
SELECT COUNT(DISTINCT nct_id) FROM a25
WHERE count=0;

 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
Done.
Done.
11977 rows affected.
1 rows affected.


[(11601,)]

16 (10 points) Rewrite the same query as immediately above, but using a different SQL mechanism (e.g. JOINs vs.  subqueries)

In [24]:
%%sql
SELECT COUNT(DISTINCT nct_id) FROM studies WHERE nct_id NOT IN (SELECT DISTINCT nct_id FROM conditions WHERE name IN (SELECT DISTINCT condition FROM factor))

 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
1 rows affected.


[(11601,)]

17 (10 points)
What are the nct_ids of other studies that have any of the conditions in study NCT00507767? Note: you may not hard-code the condition names from study NCT00507767 in your query. If there is more than one study, order by nct_id.


In [24]:
%%sql
SELECT DISTINCT l.nct_id FROM conditions l 
WHERE l.name IN ( SELECT l2.name FROM conditions l2 WHERE l2.nct_id = 'NCT00507767') AND nct_id != 'NCT00507767'


 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
4 rows affected.


[('NCT00096512',), ('NCT01528137',), ('NCT01816984',), ('NCT00939627',)]

The Jaccard Index (\url{https://en.wikipedia.org/wiki/Jaccard_index}) provides a similarity measure over sets. We can use the conditions as the set elements for calculating the Jaccard index, taking into consideration conditions in common between two studies and the total number of conditions for each study.

Note: Be careful about integer division in your calculations. If you divide two integers, you will get another integer. In this case, we want a decimal value. Consider using the NUMERIC data type when calculating the Jaccard Index.

Round all answers to 2 decimal places.

18a (10 points) What is the Jaccard Index for studies with nct_id NCT03330301 and NCT02742597? You may use the nct_ids in your query(s).



In [25]:
%%sql
DROP TABLE IF EXISTS a26 CASCADE;
CREATE TABLE a26(
    condition VARCHAR(200)
);
DROP TABLE IF EXISTS a27 CASCADE;
CREATE TABLE a27(
    condition VARCHAR(200)
);
INSERT INTO a26
SELECT name FROM conditions WHERE nct_id = 'NCT03330301';
INSERT INTO a27
SELECT name FROM conditions WHERE nct_id = 'NCT02742597';
SELECT round((SELECT CAST((SELECT COUNT(DISTINCT s.condition) FROM a26 AS s INNER JOIN a27 AS f ON s.condition = f.condition) AS numeric))/
             (SELECT CAST((SELECT COUNT(DISTINCT name) FROM conditions WHERE nct_id = 'NCT03330301' OR nct_id='NCT02742597') AS numeric)),2) AS "Jaccard Index"



 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
Done.
Done.
Done.
Done.
14 rows affected.
31 rows affected.
1 rows affected.


[(Decimal('0.07'),)]

### For the remaining questions, consider ONLY the data from studies with start\_date of 2016-06-01, and completion\_date < '2016-12-31' except where otherwise noted (e.g. specific nct_ids)


18b (no points, you need this for the next part) Compute the Jaccard Index for every pair of studies, based on conditions. 

Depending on how you write your queries, they can take a very long time to run. So, you might give some thought as to how to make them more efficient.  Note that there is a base case for computing the Jaccard Index, where if the number of conditions you are comparing from both studies is 0, the Jaccard index is defined to be 1.

You may use multiple cells to code your answer.

In [41]:
%%sql
SELECT COUNT(DISTINCT name)
FROM studies AS s JOIN conditions AS f ON s.nct_id = f.nct_id
GROUP BY s.nct_id
ORDER BY s.nct

 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
11926 rows affected.


[(37,),
 (4,),
 (36,),
 (1,),
 (1,),
 (1,),
 (1,),
 (37,),
 (2,),
 (2,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (2,),
 (1,),
 (3,),
 (3,),
 (6,),
 (1,),
 (2,),
 (1,),
 (4,),
 (1,),
 (2,),
 (1,),
 (2,),
 (46,),
 (1,),
 (1,),
 (1,),
 (1,),
 (15,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (2,),
 (2,),
 (1,),
 (2,),
 (1,),
 (2,),
 (1,),
 (1,),
 (1,),
 (3,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (36,),
 (5,),
 (1,),
 (1,),
 (3,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (2,),
 (1,),
 (3,),
 (1,),
 (1,),
 (2,),
 (1,),
 (2,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (1,),
 (1,),
 (3,),
 (1,),
 (2,),
 (1,),
 (1,),
 (1,),
 (1,),
 (3,),
 (1,),
 (1,),
 (2,),
 (2,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (3,),
 (1,),
 (1,),
 (2,),
 (1,),
 (2,),
 (2,),
 (1,),
 (2,),
 (2,),
 (1,),
 (2,),
 (1,),
 (1,),
 (1,),
 (3,),
 (2,),
 (1,),
 (1,),
 (5,),
 (2,),
 (1,),
 (11,),
 (2,),
 (1,),
 (1,),
 (4,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (2,),
 (2,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (3,),
 (2,),
 (1,),
 (2,),
 (1,),
 (1,),
 (4,),
 (4,),
 (1,),
 (2,),
 (1,),
 (1,),
 (2,),
 (1,),
 (3,),
 (3,),
 (1,),
 (1,),
 (1,),
 (5,),
 (3,),
 (2,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (2,),
 (2,),
 (1,),
 (1,),
 (1,),
 (1,),
 (7,),
 (1,),
 (1,),
 (2,),
 (5,),
 (1,),
 (1,),
 (1,),
 (5,),
 (2,),
 (1,),
 (2,),
 (4,),
 (3,),
 (1,),
 (1,),
 (1,),
 (4,),
 (1,),
 (3,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (4,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (1,),
 (3,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (2,),
 (1,),
 (6,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (3,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (3,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (1,),
 (1,),
 (4,),
 (5,),
 (1,),
 (1,),
 (3,),
 (2,),
 (3,),
 (1,),
 (3,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (5,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (3,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (9,),
 (1,),
 (3,),
 (2,),
 (2,),
 (1,),
 (1,),
 (1,),
 (2,),
 (5,),
 (1,),
 (2,),
 (1,),
 (2,),
 (1,),
 (2,),
 (2,),
 (1,),
 (2,),
 (2,),
 (2,),
 (2,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (3,),
 (4,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (2,),
 (1,),
 (1,),
 (2,),
 (1,),
 (4,),
 (2,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (4,),
 (1,),
 (2,),
 (3,),
 (2,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (4,),
 (1,),
 (2,),
 (2,),
 (2,),
 (1,),
 (2,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (8,),
 (1,),
 (1,),
 (5,),
 (1,),
 (1,),
 (1,),
 (1,),
 (41,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (4,),
 (2,),
 (1,),
 (1,),
 (1,),
 (2,),
 (3,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (3,),
 (1,),
 (2,),
 (3,),
 (2,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (3,),
 (3,),
 (3,),
 (1,),
 (1,),
 (5,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (1,),
 (4,),
 (3,),
 (1,),
 (2,),
 (1,),
 (2,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (4,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (2,),
 (1,),
 (2,),
 (2,),
 (1,),
 (1,),
 (6,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (4,),
 (1,),
 (2,),
 (3,),
 (3,),
 (1,),
 (1,),
 (4,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (2,),
 (1,),
 (1,),
 (2,),
 (2,),
 (2,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (2,),
 (2,),
 (2,),
 (6,),
 (1,),
 (1,),
 (3,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (1,),
 (8,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (2,),
 (1,),
 (2,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (2,),
 (1,),
 (2,),
 (1,),
 (1,),
 (1,),
 (1,),
 (2,),
 (1,),
 (1,),
 (2,),
 (1,),
 (2,),
 (

In [33]:
%%sql
DROP TABLE IF EXISTS a28 CASCADE;
CREATE TABLE a28(
    nct_id CHARACTER(11),
    condition VARCHAR(200)
);
INSERT INTO a28
SELECT s.nct_id,f.name
FROM studies AS s  LEFT OUTER JOIN conditions AS f ON s.nct_id = f.nct_id
WHERE s.start_date = '2016-06-01' AND s.completion_date < '2016-12-31';

DROP TABLE IF EXISTS a29 CASCADE;

CREATE TABLE a29 AS
SELECT 
s.nct_id AS nct_id1,
f.nct_id AS nct_id2,
COUNT(*) FILTER (WHERE s.condition = f.condition) AS same,
COUNT (DISTINCT s.condition) AS union1,
COUNT (DISTINCT f.condition) AS union2
FROM a28 AS s JOIN a28 AS f ON s.nct_id<f.nct_id
GROUP BY s.nct_id,f.nct_id;


ALTER TABLE a29
ADD COLUMN Jindex numeric;

UPDATE a29
SET Jindex = 1
WHERE union1+union2 = 0;

UPDATE a29
SET Jindex = CAST(same AS numeric)/(CAST(union1 AS numeric) +CAST(union2 AS numeric) -CAST(same AS numeric))
WHERE union1 + union2 != 0;

SELECT * FROM a29;


 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
Done.
Done.
239 rows affected.
Done.
17020 rows affected.
Done.
0 rows affected.
17020 rows affected.
17020 rows affected.


[('NCT01528774', 'NCT02294201', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02321995', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02376634', 0, 1, 2, Decimal('0E-20')),
 ('NCT01528774', 'NCT02386514', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02402413', 0, 1, 2, Decimal('0E-20')),
 ('NCT01528774', 'NCT02420496', 0, 1, 2, Decimal('0E-20')),
 ('NCT01528774', 'NCT02458612', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02486848', 0, 1, 2, Decimal('0E-20')),
 ('NCT01528774', 'NCT02556411', 0, 1, 2, Decimal('0E-20')),
 ('NCT01528774', 'NCT02567149', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02601859', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02645617', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02649920', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02662816', 0, 1, 2, Decimal('0E-20')),
 ('NCT01528774', 'NCT02688270', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02708004', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02712372', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02731846', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02731898', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02733978', 0, 1, 2, Decimal('0E-20')),
 ('NCT01528774', 'NCT02735525', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02737111', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02745288', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02745860', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02747823', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02753478', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02758119', 0, 1, 3, Decimal('0E-20')),
 ('NCT01528774', 'NCT02761616', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02764749', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02766283', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02766491', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02768246', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02769442', 0, 1, 2, Decimal('0E-20')),
 ('NCT01528774', 'NCT02770222', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02771080', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02771587', 0, 1, 2, Decimal('0E-20')),
 ('NCT01528774', 'NCT02773017', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02774070', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02777242', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02777827', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02780245', 0, 1, 3, Decimal('0E-20')),
 ('NCT01528774', 'NCT02780713', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02783183', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02783846', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02784847', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02784925', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02785029', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02785042', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02785094', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02785289', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02785315', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02785770', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02786082', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02786927', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02788734', 0, 1, 3, Decimal('0E-20')),
 ('NCT01528774', 'NCT02792062', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02792296', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02793063', 0, 1, 2, Decimal('0E-20')),
 ('NCT01528774', 'NCT02794038', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02797223', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02797522', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02797938', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02798289', 0, 1, 2, Decimal('0E-20')),
 ('NCT01528774', 'NCT02798393', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02798991', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02799264', 0, 1, 1, Decimal('0E-20')),
 ('NCT01528774', 'NCT02799992', 0, 1, 1,

18c (15 points) What is the average non-zero Jaccard index value in our set of studies? In other words, if you exclude pairs for which the Jaccard index value is 0, what is the average score? 

In [27]:
%%sql
SELECT CAST(AVG(Jindex) AS numeric)
FROM a29
WHERE Jindex !=0;

 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
1 rows affected.


[(Decimal('0.91666666666666666667'),)]

18d (5 points)
What percentage of study pairs have a Jaccard Index of 1?

In [28]:
%%sql
SELECT CAST((SELECT COUNT(Jindex) FROM a29  WHERE Jindex = 1)AS numeric)/CAST((SELECT COUNT(Jindex) FROM a29) AS numeric)
AS " RESULT"

 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
1 rows affected.


[(Decimal('0.00364277320799059929'),)]

## Add code to drop any custom views or new tables that you created

In [121]:
%%sql
DROP TABLE IF EXISTS a21 CASCADE;
DROP TABLE IF EXISTS a22 CASCADE;
DROP TABLE IF EXISTS a23 CASCADE;
DROP TABLE IF EXISTS a24 CASCADE;
DROP TABLE IF EXISTS a25 CASCADE;
DROP TABLE IF EXISTS a26 CASCADE;
DROP TABLE IF EXISTS a27 CASCADE;
DROP TABLE IF EXISTS a28 CASCADE;
DROP TABLE IF EXISTS a29 CASCADE;

 * postgresql+psycopg2://jf58:***@ricedb-533.ctse2r7dddre.us-east-1.rds.amazonaws.com/jf58db
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.


[]